src/main/java/io/vertx/ext/amqp/impl/AmqpConnectionImpl.java

```diff
@@ -0,0 +1,369 @@
+              return;
+            }
+
+            Map<Symbol, Object> map = new HashMap<>();
+            map.put(AmqpConnectionImpl.PRODUCT_KEY, AmqpConnectionImpl.PRODUCT);
+            if (options.getContainerId() != null) {
+              this.connection.get().setContainer(options.getContainerId());
+            }
+
+            if (options.getVirtualHost() != null) {
+              this.connection.get().setHostname(options.getVirtualHost());
+            }
+
+            this.connection.get()
+              .setProperties(map)
+              .disconnectHandler(ignored -> {
+                onEnd();
+                closed.set(true);
+              })
+              .closeHandler(ignored -> {
+                try {
+                  onDisconnect();
+                } finally {
+                  onEnd();
+                  closed.set(true);
+                }
+              })
+              .openHandler(conn -> {
+                if (conn.succeeded()) {
+                  client.register(this);
+                  closed.set(false);
+                  connectionHandler.handle(Future.succeededFuture(this));
+                } else {
+                  closed.set(true);
+                  connectionHandler.handle(conn.mapEmpty());
+                }
+              });
+
+            this.connection.get().open();
+          } else {
+            connectionHandler.handle(ar.mapEmpty());
+          }
+        });
+  }
+
+  /**
+   * Must be called on context.
+   */
+  private void onDisconnect() {
+    ProtonConnection conn = connection.getAndSet(null);
+    if (conn != null) {
+      try {
+        conn.close();
+      } finally {
+        conn.disconnect();
+      }
+    }
+  }
+
+  /**
+   * Must be called on context.
+   */
+  private void onEnd() {
+    Handler<Void> handler;
+    synchronized (this) {
+      handler = endHandler;
+      endHandler = null;
+    }
+    if (handler != null && !closed.get()) {
+      handler.handle(null);
+    }
+  }
+
+  void runOnContext(Handler<Void> action) {
+    context.runOnContext(action);
+  }
+
+  void runWithTrampoline(Handler<Void> action) {
+    if (Vertx.currentContext() == context) {
+      action.handle(null);
+    } else {
+      runOnContext(action);
+    }
+  }
+
+  /**
+   * Must be called on context.
+   */
+  private boolean isLocalOpen() {
+    ProtonConnection conn = this.connection.get();
+    return conn != null
+      && ((ProtonConnectionImpl) conn).getLocalState() == EndpointState.ACTIVE;
+  }
+
+  /**
+   * Must be called on context.
+   */
+  private boolean isRemoteOpen() {
+    ProtonConnection conn = this.connection.get();
+    return conn != null
+      && ((ProtonConnectionImpl) conn).getRemoteState() == EndpointState.ACTIVE;
+  }
+
+  @Override
+  public synchronized AmqpConnection endHandler(Handler<Void> endHandler) {
+    this.endHandler = endHandler;
+    return this;
+  }
+
+  @Override
+  public AmqpConnection close(Handler<AsyncResult<Void>> done) {
```

Comment:
The close(..) process still isnt going to cater for a disconnection-without-close-response occurring after it starts.

In that case the handler passed to close(..) wont fire since no close frame arrives. Neither will the closeHandler(..) be fired for the same reason, but also that the underlying connections handler which did that work was replaced during close(..) and no longer does. The endHandler also seems like it won't be called in this scenario since 'closed' is set true during close(..) and the endHandler firing is gated on that. The last bit actually makes sense, but expectation is another handler is being called instead and it wont be here. Close(..) needs to make use of the disconnect handler. At least one of these 3 handlers should be called in that scenario.

Going back to an earlier comment from @vietj  about the closeHandler and endHandler both existing, I do now think change is needed there. Having a closeHandler(<handler>), endHandler(<handler>), and close(<handler>) is rather duplicative, I hadn't been taking note of close(..) having its own handler already, so if its going to then I also think one of the others should go away and their functionality be combined (i.e a handler that is called if either a close frame arrives or the connection disconnects without one, and close has not been called [with non-null handler?].

Reply:
I don't understand how to handle this case, would this be enough:
```
this.connection.get().closeHandler(pc -> {
      if (remoteCloseHandler != null) {
        runWithTrampoline(x -> {
          try {
            onDisconnect();
          } finally {
            onEnd();
            closed.set(true);
          }
          remoteCloseHandler.handle(this);
        });
      }
    });
```

Or should I also add a `disconnectHandler`:

```
Handler<AsyncResult<Void>> handler = pc -> {
      if (remoteCloseHandler != null) {
        runWithTrampoline(x -> {
          try {
            onDisconnect();
          } finally {
            onEnd();
            closed.set(true);
          }
          remoteCloseHandler.handle(this);
        });
      }
    };
    this.connection.get()
      .closeHandler(x -> handler.handle(x.mapEmpty()))
      .disconnectHandler(x -> handler.handle(Future.succeededFuture()));
```

You mention 3 handlers, but I only have disconnect and close. What's the third signals?

About the last part, that would mean removing the `endHandler`.

Reply:
The three handlers noted are from AmqpConnection:
AmqpConnection#close(\<handler\>);
AmqpConnection#closeHandler(\<handler\>);
AmqpConnection#endHandler(\<handler\>);

If the intent is to fire the AmqpConnection#close(\<handler\>) once that method is called, I think one of the other handlers goes away.

I think you should add a disconnectHandler as well as closeHandler to the underlying ProtonConnection within the AmqpConnection#close(\<handler\>) method impl. That way regardless whether the process completes normally, or if the underlying connection disconnects without a close frame being received, you get notified and can complete the handler passed to AmqpConnection#close(\<handler\>). The AmqpConnection#closeHandler(\<handler\>) and AmqpConnection#endHandler(\<handler\>) handlers can be documented not to fire after the point AmqpConnection#close(\<handler\>) returns. At which point removing one of them seems to make sense, and youve already somewhat started that by making AmqpConnection#closeHandler(\<handler\>) fire for disconnects as well as close frames that occur before AmqpConnection#close(\<handler\>).

Note that the underlying ProtonConnection closeHandler and disconnectHandler can/do both fire though, so you should guard completions etc on that. The former is purely indicating receipt of a close frame (either a response to locally-initiated close, or a remotely-initiated close), the latter the TCP disconnection, even if we asked for the closure. So its typically close+disconnect if a close frame is received, or just disconnect if a close frame is not received. I think a AmqpConnection#closeHandler(\<handler\>) being registered after the latest change would currently barf a little upon a server-initiated close.

Reply:
I've added an `exceptionHandler` replacing the `onEnd`(more inlined with the rest of Vert.x).
This handler is called on disconnect and unexpected closing (the server closes the connection).

In `close`, I register both `onDisconnect` and `onClose` handling the disconnect and close situation.